  int32 cmd_code
  int32 cmd_value
  int32 type